include/dd/ComplexTable.hpp

```diff
@@ -380,52 +416,54 @@ namespace dd {
                << ", gc calls: " << gcCalls
                << ", gc runs: " << gcRuns
                << "\n";
-            //clang-format on
+            // clang-format on
             return os;
         }
 
-        // table lookup statistics
-        std::size_t collisions = 0;
-        std::size_t insert_collisions = 0;
-        std::size_t hits = 0;
-        std::size_t find_or_inserts = 0;
-        std::size_t lookups = 0;
-        std::size_t inserts = 0;
-        std::size_t lower_neighbors = 0;
-        std::size_t upper_neighbors = 0;
-
     private:
-        using Bucket = Entry *;
-        using Table = std::array<Bucket, NBUCKET>;
+        using Bucket = Entry*;
+        using Table  = std::array<Bucket, NBUCKET>;
 
         Table table{};
 
+        std::array<Entry*, NBUCKET> last_entry_table{};
+
+        // table lookup statistics
+        std::size_t collisions        = 0;
+        std::size_t insert_collisions = 0;
+        std::size_t hits              = 0;
+        std::size_t find_or_inserts   = 0;
+        std::size_t lookups           = 0;
+        std::size_t inserts           = 0;
+        std::size_t lower_neighbors   = 0;
+        std::size_t upper_neighbors   = 0;
+
         // numerical tolerance to be used for floating point values
         static inline fp TOLERANCE = 1e-13;
 
-        Entry *available{};
-        std::vector<std::vector<Entry>> chunks{};
-        std::size_t chunkID;
+        Entry*                                available{};
+        std::vector<std::vector<Entry>>       chunks{};
+        std::size_t                           chunkID;
         typename std::vector<Entry>::iterator chunkIt;
         typename std::vector<Entry>::iterator chunkEndIt;
-        std::size_t allocationSize;
+        std::size_t                           allocationSize;
 
         std::size_t allocations = 0;
-        std::size_t count = 0;
-        std::size_t peakCount = 0;
+        std::size_t count       = 0;
+        std::size_t peakCount   = 0;
 
         // garbage collection
         std::size_t gcCalls = 0;
-        std::size_t gcRuns = 0;
+        std::size_t gcRuns  = 0;
         std::size_t gcLimit = 100000;
 
-        inline Entry *find_or_insert(const std::size_t key, const fp val) {
-            const fp val_tol = val - TOLERANCE;
+        inline Entry* find_or_insert(const std::size_t key, const fp val) {
```

Comment:
```suggestion
        inline Entry* findOrInsert(const std::size_t key, const fp val) {
```
is there a better name for this? I'd say `insert` itself would already be enough, but this collides with the other `insert` method.

Reply:
I think the same suits the function well as it primarily tries to find elements and only inserts nothings found (alluding to short-circuit evaluation)

Reply:
In C++ (and I would assume other languages as well) `std::set::insert` does precisely this.
It inserts the value if and only if it is not already found in the container, returning an iterator to the found (or inserted) element.
https://www.cplusplus.com/reference/set/set/insert/

Reply:
On the other hand, `std::set` does not have functions to insert regardless of previous existence. I'd keep the name.

Reply:
Fine 👍🏼